dot/state/block.go

```diff
@@ -765,3 +798,99 @@ func (bs *BlockState) setArrivalTime(hash common.Hash, arrivalTime time.Time) er
 	binary.LittleEndian.PutUint64(buf, uint64(arrivalTime.UnixNano()))
 	return bs.db.Put(arrivalTimeKey(hash), buf)
 }
+
+// HandleRuntimeChanges handles the update in runtime.
+func (bs *BlockState) HandleRuntimeChanges(newState *rtstorage.TrieState, rt runtime.Instance, bHash common.Hash) error {
+	currCodeHash, err := newState.LoadCodeHash()
+	if err != nil {
+		return err
+	}
+
+	codeHash := rt.GetCodeHash()
+	if bytes.Equal(codeHash[:], currCodeHash[:]) {
+		bs.StoreRuntime(bHash, rt)
+		return err
+	}
+
+	logger.Info("🔄 detected runtime code change, upgrading...", "block", bHash, "previous code hash", codeHash, "new code hash", currCodeHash)
+	code := newState.LoadCode()
+	if len(code) == 0 {
+		return errors.New("new :code is empty")
+	}
+
+	codeSubBlockHash := bs.baseState.LoadCodeSubstitutedBlockHash()
+	if !codeSubBlockHash.Equal(common.Hash{}) {
+		newVersion, err := rt.CheckRuntimeVersion(code) //nolint
+		if err != nil {
+			return err
+		}
+
+		previousVersion, _ := rt.Version()
+		if previousVersion.SpecVersion() == newVersion.SpecVersion() {
+			return nil
+		}
+
+		logger.Info("🔄 detected runtime code change, upgrading...", "block", bHash,
+			"previous code hash", codeHash, "new code hash", currCodeHash,
+			"previous spec version", previousVersion.SpecVersion(), "new spec version", newVersion.SpecVersion())
+	}
+
+	rtCfg := &wasmer.Config{
+		Imports: wasmer.ImportsNodeRuntime,
+	}
+
+	localStorage, err := utils.SetupDatabase(filepath.Join(bs.dbPath, bHash.String(), "local_storage"), true)
+	if err != nil {
+		return err
+	}
+
+	ns := rt.NodeStorage()
+	ns.LocalStorage = localStorage
+
+	rtCfg.Storage = newState
+	rtCfg.Keystore = rt.Keystore()
+	rtCfg.NodeStorage = ns
+	rtCfg.Network = rt.NetworkService()
+	rtCfg.CodeHash = currCodeHash
+
+	if rt.Validator() {
+		rtCfg.Role = 4
+	}
+
+	instance, err := wasmer.NewInstance(code, rtCfg)
+	if err != nil {
+		return err
+	}
+
+	bs.StoreRuntime(bHash, instance)
+
+	err = bs.baseState.StoreCodeSubstitutedBlockHash(common.Hash{})
+	if err != nil {
+		return fmt.Errorf("failed to update code substituted block hash: %w", err)
+	}
+
+	return nil
+}
+
+// GetRuntime gets the runtime for the corresponding block hash.
+func (bs *BlockState) GetRuntime(hash *common.Hash) (runtime.Instance, error) {
+	if hash == nil {
+		rt, err := bs.bt.GetBlockRuntime(bs.BestBlockHash())
+		if err != nil {
+			return nil, err
+		}
+		return rt, nil
+	}
+
+	return bs.bt.GetBlockRuntime(*hash)
+}
+
+// StoreRuntime stores the runtime for corresponding block hash.
+func (bs *BlockState) StoreRuntime(hash common.Hash, rt runtime.Instance) {
+	bs.bt.StoreRuntime(hash, rt)
+}
+
+// GetAllBlocks get all the blocks
+func (bs *BlockState) GetAllBlocks() []common.Hash {
+	return bs.bt.GetAllBlocks()
+}
```

Comment:
maybe name `GetAllUnfinalisedBlocks` for clarity?

Reply:
Done